### ISSUE TITLE
SIM300: CONSTANT_CASE variables are improperly flagged for yoda violation

### ISSUE THREAD

**Issue by rseeley** [astral-sh/ruff#6956]
Ruff version: 0.0.286.

Values with CONSTANT_CASE seem to be flagged improperly by the [yoda rule (SIM300)](https://beta.ruff.rs/docs/rules/yoda-conditions/).
```python
UPPER_VAL = 1

assert UPPER_VAL == 1

UPPER_LIST = ['upper']

assert UPPER_LIST == ['upper']

lower_val = 1

assert lower_val == 1
```

Running `ruff . --select SIM300 --isolated` gives the following result:
```
path/to/some/test_file.py:45:12: SIM300 [*] Yoda conditions are discouraged, use `['upper'] == UPPER_LIST` instead
Found 1 error.
```
I may be incorrect, but it seems like `assert UPPER_LIST == ['upper']` should be correct.

Though not shown in my example above, I first noticed this when using a constant-cased class attribute, in case that helps with testing a fix.

**Comment by charliermarsh** [comment]
Hmm, I think we're not recognizing `['upper']` as a "constant".

**Comment by rseeley** [comment]
Ah right, I meant to mention it only seems to happen with mutable data types.

**Comment by tdulcet** [comment]
Doing simple calculations with constant values also triggers this:
```py
primes_diff = [
    # Large list...
]
PRIMES_PTAB_ENTRIES = len(primes_diff)
assert PRIMES_PTAB_ENTRIES == 6542 - 1
```
Ruff output:
```
$ ruff --select SIM300 --isolated factor.py
factor.py:260:8: SIM300 [*] Yoda conditions are discouraged, use `6542 - 1 == PRIMES_PTAB_ENTRIES` instead
Found 1 error.
[*] 1 potentially fixable with the --fix option.
```
There is a [full example here](https://gist.github.com/tdulcet/0c4567c0c295025723807026a4d27d33/#file-factor-py-L95).

**Comment by charliermarsh** [comment]
This came up again in https://github.com/astral-sh/ruff/issues/8759.

**Comment by ThiefMaster** [comment]
> We should probably flag `assert 1 == DummyHandler.CONFIG`, which we don't today -- we consider both of those expressions to be equally "constant".

Maybe `is_constant_like` could return a numeric value (False -> 0, uppercase -> 1, literal -> 2) so an actual literal can get higher priority than a constant in that check?

Looking at the example from @tdulcet, I guess it'd also need to consider arithmetic operations on literals to still be a literal
